lib/propolis/src/block/attachment.rs

```diff
@@ -0,0 +1,362 @@
+    pub fn pause(&self) {
+        self.0.access_fallback(
+            |data, state| {
+                state.pause();
+                data.notify();
+            },
+            |state| state.pause(),
+        );
+    }
+
+    /// Clear the paused state on this device, allowing the backend (if
+    /// attached) to retrieve requests once again.
+    pub fn resume(&self) {
+        self.0.access_fallback(
+            |data, state| {
+                state.resume();
+                data.notify();
+            },
+            |state| state.resume(),
+        );
+    }
+
+    /// Attempt to detach device from backend
+    pub fn detach(&self) -> Result<(), attachment::DetachError> {
+        if let Some((dev, backend)) = self
+            .0
+            .access(|data, _state| (data.device.clone(), data.backend.clone()))
+        {
+            detach(dev, backend)
+        } else {
+            Err(attachment::DetachError::FrontNotAttached)
+        }
+    }
+}
+
+/// Reason why next request is unavailable from associated device
+pub enum ReqError {
+    /// No request is pending from the device
+    NonePending,
+    /// Processing of requests from the device is paused
+    Paused,
+    /// Backend is not attached to any device
+    Detached,
+    /// Backend is stopping workers
+    Stopped,
+}
+
+pub struct BackendAttachment(attachment::BackAttachment<Arc<BlockData>>);
+impl BackendAttachment {
+    pub fn new() -> Self {
+        Self(attachment::BackAttachment::new())
+    }
+
+    /// Attempt to retrieve the next [`Request`] from the attached (if any)
+    /// device.
+    ///
+    /// Will return an error if:
+    ///
+    /// - No device is attached
+    /// - The device is paused
+    /// - The backend is stopped
+    /// - No requests are queued in the device
+    pub fn next_req(&self) -> Result<Request, ReqError> {
+        match self.0.access(|data, state| data.next_req(state)) {
+            None => Err(ReqError::Detached),
+            Some(Err(e)) => Err(e),
+            Some(Ok(r)) => Ok(r),
+        }
+    }
+
+    /// Block (synchronously) in order to retrieve the next [`Request`] from the
+    /// device.  Will return [`None`] if no device is attached, or the backend
+    /// is stopped, otherwise it will block until a request is available.
+    pub fn block_for_req(&self) -> Option<Request> {
+        match self.0.access(|data, state| {
+            // If we do not immediately get a request, clone the BackendData
+            // so we can do our blocking through that, instead of repeated calls
+            // into BackAttachment::access()
+            data.next_req(state).map_err(|_e| (data.clone(), state.clone()))
+        })? {
+            Ok(req) => Some(req),
+            Err((be, state)) => be.block_for_req(&state),
+        }
+    }
+
+    /// Get a [Waiter], through which one can asynchronously poll for new
+    /// requests via [`Waiter::for_req()`].  Will return [`None`] if no device
```

Comment:
Very silly nit: inconsistent use of backticks in the links here:
```suggestion
    /// Get a [`Waiter`], through which one can asynchronously poll for new
    /// requests via [`Waiter::for_req()`].  Will return [`None`] if no device
```

Reply:
I'm leaving them around the `for_req` references, because otherwise the nvim-lsp renderer complains (mildly) about the markdown formatting and the underscore.  Otherwise I'll clean up the rest.